### DISCUSSION TITLE
Do you want a new online editor?

### DISCUSSION THREAD

**Discussion by gera2ld** [violentmonkey/violentmonkey#1753] [Ideas]
![image](https://user-images.githubusercontent.com/3139113/218504113-0999320c-baca-4b97-b408-f2f111765f6d.png)

Just an idea, it can be

- powerful, based on Monaco editor, with great TypeScript support
- easy to integrate, with simple and intuitive APIs to interact with Violentmonkey
- hassle-free, always enjoy the latest features

I think it's much better than tweaking the CodeMirror editor inside Violentmonkey. So why not keep the internal editor as simple as possible and build a powerful one in the free world?

[👉 Play](https://dweb.link/ipns/vm.gera2ld.space)

**Comment by tophf** [comment]
The built-in editor is already usable for trivial editing tasks, no need to complicate it, indeed, but it's important to make the online editor fully cacheable so it can work offline. AFAIK, it is achievable via a service worker and optionally a PWA manifest.

It'd be also good to allow using it inside an iframe so Violentmonkey can embed it seamlessly in the UI.

**Reply by gera2ld** [reply]
Yes, it's totally feasible to build a PWA and work offline.
I'm afraid there will be review problems if we embed it into the extension.

**Reply by tophf** [reply]
Extensions can put any external content inside an iframe in their UI, and many of them do, AFAIK.

**Reply by tophf** [reply]
Ah, there's a [sandbox](https://developer.chrome.com/extensions/manifest/sandbox) feature in manifest.json that allows including external scripts, so this is conceptually allowed inside extensions, although we probably won't even need sandboxing because the editor will be fully encapsulated within its origin, so there'll be nothing unsafe inside the extension's origin.

**Reply by tophf** [reply]
Tampermonkey implemented it:

> userscript editing at [vscode.dev](https://vscode.dev/?connectTo=tampermonkey) if the [Tampermonkey Editors extension](https://chrome.google.com/webstore/detail/lieodnapokbjkkdkhdljlllmgkmdokcm) is installed

We can probably do the same but without the need for a separate extension.
